File: driftwatch/backend/internal/api/oauth.go
package api

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/golang-jwt/jwt/v5"
	"github.com/google/uuid"

	"github.com/YOURUSERNAME/driftwatch/internal/crypto"
	"github.com/YOURUSERNAME/driftwatch/internal/db"
)

// OAuthConfig holds the GitHub OAuth app credentials and the dashboard URL we
// redirect back to after login.
type OAuthConfig struct {
	ClientID     string
	ClientSecret string
	// DashboardURL is where the callback sends the browser with the issued JWT,
	// e.g. https://driftwatch.pages.dev. The token is appended as #token=...
	DashboardURL string
	// BackendURL, when set, is this service's external base URL used to build the
	// OAuth redirect_uri (e.g. https://driftwatch-3drv.onrender.com). If empty it
	// is reconstructed from the request.
	BackendURL string
}

// Enabled reports whether GitHub OAuth is configured.
func (o OAuthConfig) Enabled() bool {
	return o.ClientID != "" && o.ClientSecret != ""
}

const githubAuthorizeURL = "https://github.com/login/oauth/authorize"
const githubTokenURL = "https://github.com/login/oauth/access_token"
const githubUserURL = "https://api.github.com/user"
const githubEmailsURL = "https://api.github.com/user/emails"

// handleGithubLogin redirects the browser to GitHub's consent screen. We request
// `repo` and `read:user` scopes so the same token can later read the user's
// private compose files.
func (a *API) handleGithubLogin(c *gin.Context) {
	if !a.oauth.Enabled() {
		respondError(c, http.StatusServiceUnavailable, "GitHub login is not configured", "OAUTH_DISABLED")
		return
	}

	// CSRF state: random token round-tripped via a short-lived cookie.
	state := randomToken()
	c.SetCookie("dw_oauth_state", state, 600, "/", "", true, true)

	q := url.Values{}
	q.Set("client_id", a.oauth.ClientID)
	q.Set("redirect_uri", a.backendBaseURL(c)+"/api/auth/github/callback")
	q.Set("scope", "read:user user:email repo")
	q.Set("state", state)
	q.Set("allow_signup", "true")

	c.Redirect(http.StatusFound, githubAuthorizeURL+"?"+q.Encode())
}

// handleGithubCallback completes the OAuth dance: verify state, exchange the
// code for an access token, fetch the GitHub profile, upsert the user, issue our
// own JWT, and redirect to the dashboard with the token in the URL fragment.
func (a *API) handleGithubCallback(c *gin.Context) {
	if !a.oauth.Enabled() {
		respondError(c, http.StatusServiceUnavailable, "GitHub login is not configured", "OAUTH_DISABLED")
		return
	}

	// Verify CSRF state.
	wantState, _ := c.Cookie("dw_oauth_state")
	gotState := c.Query("state")
	if wantState == "" || gotState == "" || wantState != gotState {
		a.redirectToDashboardError(c, "state_mismatch")
		return
	}
	c.SetCookie("dw_oauth_state", "", -1, "/", "", true, true)

	code := c.Query("code")
	if code == "" {
		a.redirectToDashboardError(c, "missing_code")
		return
	}

	ctx, cancel := context.WithTimeout(c.Request.Context(), 15*time.Second)
	defer cancel()

	accessToken, err := a.exchangeGithubCode(ctx, code, a.backendBaseURL(c)+"/api/auth/github/callback")
	if err != nil {
		a.redirectToDashboardError(c, "token_exchange_failed")
		return
	}

	profile, err := fetchGithubProfile(ctx, accessToken)
	if err != nil {
		a.redirectToDashboardError(c, "profile_fetch_failed")
		return
	}

	// Encrypt the OAuth token before storing — it's reused for private repos.
	encToken, err := crypto.Encrypt(accessToken)
	if err != nil {
		a.redirectToDashboardError(c, "encrypt_failed")
		return
	}

	user, err := a.queries.UpsertGithubUser(ctx, db.UpsertGithubUserParams{
		Email:                profile.Email,
		GithubID:             profile.ID,
		GithubLogin:          profile.Login,
		AvatarURL:            profile.AvatarURL,
		GithubTokenEncrypted: encToken,
	})
	if err != nil {
		a.redirectToDashboardError(c, "user_upsert_failed")
		return
	}

	signed, err := a.issueJWT(user.ID)
	if err != nil {
		a.redirectToDashboardError(c, "token_sign_failed")
		return
	}

	// Hand the JWT to the SPA via the URL fragment (never sent to a server, not
	// logged). The dashboard's /auth/callback page reads it and stores it.
	dest := strings.TrimRight(a.oauth.DashboardURL, "/") + "/auth/callback#token=" + url.QueryEscape(signed)
	c.Redirect(http.StatusFound, dest)
}

// issueJWT mints a 24h DriftWatch session token for a user id.
func (a *API) issueJWT(userID uuid.UUID) (string, error) {
	now := time.Now()
	claims := Claims{
		UserID: userID,
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(now.Add(24 * time.Hour)),
			IssuedAt:  jwt.NewNumericDate(now),
			NotBefore: jwt.NewNumericDate(now),
			Issuer:    "driftwatch",
		},
	}
	return jwt.NewWithClaims(jwt.SigningMethodHS256, claims).SignedString(a.jwtSecret)
}

func (a *API) exchangeGithubCode(ctx context.Context, code, redirectURI string) (string, error) {
	form := url.Values{}
	form.Set("client_id", a.oauth.ClientID)
	form.Set("client_secret", a.oauth.ClientSecret)
	form.Set("code", code)
	form.Set("redirect_uri", redirectURI)

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, githubTokenURL, strings.NewReader(form.Encode()))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<16))

	var out struct {
		AccessToken string `json:"access_token"`
		Error       string `json:"error"`
	}
	if err := json.Unmarshal(body, &out); err != nil {
		return "", err
	}
	if out.AccessToken == "" {
		return "", fmt.Errorf("github: no access token (%s)", out.Error)
	}
	return out.AccessToken, nil
}

type githubProfile struct {
	ID        int64  `json:"id"`
	Login     string `json:"login"`
	AvatarURL string `json:"avatar_url"`
	Email     string `json:"email"`
}

func fetchGithubProfile(ctx context.Context, accessToken string) (githubProfile, error) {
	var p githubProfile
	if err := githubGet(ctx, githubUserURL, accessToken, &p); err != nil {
		return p, err
	}
	// The /user email is null when the user keeps it private; fall back to the
	// primary verified email from /user/emails.
	if p.Email == "" {
		var emails []struct {
			Email    string `json:"email"`
			Primary  bool   `json:"primary"`
			Verified bool   `json:"verified"`
		}
		if err := githubGet(ctx, githubEmailsURL, accessToken, &emails); err == nil {
			for _, e := range emails {
				if e.Primary && e.Verified {
					p.Email = e.Email
					break
				}
			}
		}
	}
	return p, nil
}

func githubGet(ctx context.Context, urlStr, accessToken string, out interface{}) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, urlStr, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+accessToken)
	req.Header.Set("Accept", "application/vnd.github+json")
	req.Header.Set("User-Agent", "DriftWatch")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("github GET %s: status %d", urlStr, resp.StatusCode)
	}
	body, err := io.ReadAll(io.LimitReader(resp.Body, 1<<18))
	if err != nil {
		return err
	}
	return json.Unmarshal(body, out)
}

// redirectToDashboardError sends the browser back to the dashboard login with an
// error code in the query string, so the SPA can show a toast.
func (a *API) redirectToDashboardError(c *gin.Context, code string) {
	dest := strings.TrimRight(a.oauth.DashboardURL, "/") + "/login?error=" + url.QueryEscape(code)
	c.Redirect(http.StatusFound, dest)
}

// backendBaseURL reconstructs this backend's external base URL from the request
// (honoring proxies via X-Forwarded-Proto), used to build the OAuth redirect_uri.
func (a *API) backendBaseURL(c *gin.Context) string {
	if a.oauth.BackendURL != "" {
		return strings.TrimRight(a.oauth.BackendURL, "/")
	}
	scheme := "https"
	if proto := c.GetHeader("X-Forwarded-Proto"); proto != "" {
		scheme = proto
	} else if c.Request.TLS == nil {
		scheme = "http"
	}
	return scheme + "://" + c.Request.Host
}

func randomToken() string {
	b := make([]byte, 16)
	_, _ = rand.Read(b)
	return hex.EncodeToString(b)
}

File: driftwatch/backend/src/api/routes/oauth.ts
import { randomBytes } from 'node:crypto';

import { Router } from 'express';
import type { Request, Response } from 'express';

import { upsertGithubUser } from '../../db/users.js';
import { oauthEnabled } from '../../env.js';
import { encrypt } from '../../services/crypto.js';
import { issueJWT } from '../auth.js';
import type { ApiDeps } from '../deps.js';
import { respondError } from '../respond.js';

const GITHUB_AUTHORIZE_URL = 'https://github.com/login/oauth/authorize';
const GITHUB_TOKEN_URL = 'https://github.com/login/oauth/access_token';
const GITHUB_USER_URL = 'https://api.github.com/user';
const GITHUB_EMAILS_URL = 'https://api.github.com/user/emails';

const STATE_COOKIE = 'dw_oauth_state';
const STATE_TTL_SECONDS = 600;
const EXCHANGE_TIMEOUT_MS = 15_000;

const COOKIE_OPTS = {
	path: '/',
	httpOnly: true,
	secure: true,
	sameSite: 'lax'
} as const;

/**
 * GitHub OAuth sign-in: the browser hits /login, is bounced to GitHub, and the
 * /callback issues a DriftWatch JWT and redirects back to the dashboard.
 */
export function oauthRoutes(deps: ApiDeps): Router {
	const r = Router();

	/**
	 * Redirects to GitHub's consent screen. We request `repo` and `read:user`
	 * so the same token can later read the user's private compose files.
	 */
	r.get('/auth/github/login', (req, res) => {
		if (!oauthEnabled(deps.oauth)) {
			respondError(res, 503, 'GitHub login is not configured', 'OAUTH_DISABLED');
			return;
		}

		// CSRF state: a random token round-tripped via a short-lived cookie.
		const state = randomBytes(16).toString('hex');
		res.cookie(STATE_COOKIE, state, {
			...COOKIE_OPTS,
			maxAge: STATE_TTL_SECONDS * 1000
		});

		const q = new URLSearchParams({
			client_id: deps.oauth.clientId,
			redirect_uri: backendBaseURL(deps, req) + '/api/auth/github/callback',
			scope: 'read:user user:email repo',
			state,
			allow_signup: 'true'
		});

		res.redirect(302, `${GITHUB_AUTHORIZE_URL}?${q.toString()}`);
	});

	/**
	 * Completes the OAuth dance: verify state, exchange the code for an access
	 * token, fetch the GitHub profile, upsert the user, issue our own JWT, and
	 * redirect to the dashboard with the token in the URL fragment.
	 */
	r.get('/auth/github/callback', async (req, res) => {
		if (!oauthEnabled(deps.oauth)) {
			respondError(res, 503, 'GitHub login is not configured', 'OAUTH_DISABLED');
			return;
		}

		const wantState = req.cookies?.[STATE_COOKIE] as string | undefined;
		const gotState = typeof req.query.state === 'string' ? req.query.state : '';
		if (!wantState || !gotState || wantState !== gotState) {
			redirectToDashboardError(deps, res, 'state_mismatch');
			return;
		}
		res.clearCookie(STATE_COOKIE, COOKIE_OPTS);

		const code = typeof req.query.code === 'string' ? req.query.code : '';
		if (!code) {
			redirectToDashboardError(deps, res, 'missing_code');
			return;
		}

		let accessToken: string;
		try {
			accessToken = await exchangeGithubCode(
				deps,
				code,
				backendBaseURL(deps, req) + '/api/auth/github/callback'
			);
		} catch {
			redirectToDashboardError(deps, res, 'token_exchange_failed');
			return;
		}

		let profile: GithubProfile;
		try {
			profile = await fetchGithubProfile(accessToken);
		} catch {
			redirectToDashboardError(deps, res, 'profile_fetch_failed');
			return;
		}

		// Encrypt the OAuth token before storing — it's reused for private repos.
		let encToken: string;
		try {
			encToken = encrypt(accessToken);
		} catch {
			redirectToDashboardError(deps, res, 'encrypt_failed');
			return;
		}

		let user;
		try {
			user = await upsertGithubUser(deps.db, {
				email: profile.email,
				githubId: profile.id,
				githubLogin: profile.login,
				avatarUrl: profile.avatar_url,
				githubTokenEncrypted: encToken
			});
		} catch {
			redirectToDashboardError(deps, res, 'user_upsert_failed');
			return;
		}

		let token: string;
		try {
			token = issueJWT(user.id, deps.jwtSecret).token;
		} catch {
			redirectToDashboardError(deps, res, 'token_sign_failed');
			return;
		}

		// Hand the JWT to the SPA via the URL fragment (never sent to a server,
		// not logged). The dashboard's /auth/callback page reads and stores it.
		const dest =
			trimTrailingSlash(deps.oauth.dashboardUrl) +
			'/auth/callback#token=' +
			encodeURIComponent(token);
		res.redirect(302, dest);
	});

	return r;
}

async function exchangeGithubCode(
	deps: ApiDeps,
	code: string,
	redirectURI: string
): Promise<string> {
	const form = new URLSearchParams({
		client_id: deps.oauth.clientId,
		client_secret: deps.oauth.clientSecret,
		code,
		redirect_uri: redirectURI
	});

	const res = await fetch(GITHUB_TOKEN_URL, {
		method: 'POST',
		headers: {
			'Content-Type': 'application/x-www-form-urlencoded',
			Accept: 'application/json'
		},
		body: form.toString(),
		signal: AbortSignal.timeout(EXCHANGE_TIMEOUT_MS)
	});

	const out = (await res.json()) as { access_token?: string; error?: string };
	if (!out.access_token) {
		throw new Error(`github: no access token (${out.error ?? ''})`);
	}
	return out.access_token;
}

interface GithubProfile {
	id: number;
	login: string;
	avatar_url: string;
	email: string;
}

async function fetchGithubProfile(accessToken: string): Promise<GithubProfile> {
	const raw = await githubGet<Partial<GithubProfile>>(GITHUB_USER_URL, accessToken);
	const profile: GithubProfile = {
		id: raw.id ?? 0,
		login: raw.login ?? '',
		avatar_url: raw.avatar_url ?? '',
		email: raw.email ?? ''
	};

	// The /user email is null when the user keeps it private; fall back to the
	// primary verified address from /user/emails.
	if (profile.email === '') {
		try {
			const emails = await githubGet<
				{ email: string; primary: boolean; verified: boolean }[]
			>(GITHUB_EMAILS_URL, accessToken);
			const primary = emails.find((e) => e.primary && e.verified);
			if (primary) profile.email = primary.email;
		} catch {
			// Leave the email empty; the schema allows it.
		}
	}

	return profile;
}

async function githubGet<T>(url: string, accessToken: string): Promise<T> {
	const res = await fetch(url, {
		headers: {
			Authorization: `Bearer ${accessToken}`,
			Accept: 'application/vnd.github+json',
			'User-Agent': 'DriftWatch'
		},
		signal: AbortSignal.timeout(EXCHANGE_TIMEOUT_MS)
	});
	if (!res.ok) {
		throw new Error(`github GET ${url}: status ${res.status}`);
	}
	return (await res.json()) as T;
}

/**
 * Sends the browser back to the dashboard login with an error code in the query
 * string, so the SPA can show a toast.
 */
function redirectToDashboardError(deps: ApiDeps, res: Response, code: string): void {
	const dest =
		trimTrailingSlash(deps.oauth.dashboardUrl) +
		'/login?error=' +
		encodeURIComponent(code);
	res.redirect(302, dest);
}

/**
 * Reconstructs this backend's external base URL from the request (honoring
 * proxies via X-Forwarded-Proto), used to build the OAuth redirect_uri.
 */
export function backendBaseURL(deps: ApiDeps, req: Request): string {
	if (deps.oauth.backendUrl !== '') return trimTrailingSlash(deps.oauth.backendUrl);

	const forwarded = req.get('X-Forwarded-Proto');
	const scheme = forwarded
		? forwarded
		: (req.socket as { encrypted?: boolean }).encrypted
			? 'https'
			: 'http';
	return `${scheme}://${req.get('host') ?? ''}`;
}

function trimTrailingSlash(s: string): string {
	return s.replace(/\/+$/, '');
}
